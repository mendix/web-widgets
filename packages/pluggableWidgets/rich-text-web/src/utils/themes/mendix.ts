import Quill from "quill";

import SnowTheme from "quill/themes/snow";

class MendixTheme extends SnowTheme {
    buildPickers(selects: NodeListOf<HTMLSelectElement>, icons: Record<string, string | Record<string, string>>): void {
        super.buildPickers(selects, icons);
    }
}

Quill.register({ "themes/snow": MendixTheme }, true);
