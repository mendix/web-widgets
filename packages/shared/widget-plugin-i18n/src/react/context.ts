import { createContext } from "react";
import { PluginI18n } from "../types";

const contextPath = "com.mendix.widget.plugin.i18n.context" as const;

type I18nContext = React.Context<PluginI18n | null>;

const createI18nContext = (): I18nContext => createContext<PluginI18n | null>(null);

declare global {
    interface Window {
        [contextPath]?: ReturnType<typeof createI18nContext>;
    }
}

export function getGlobalI18nPluginContext(): I18nContext {
    return window[contextPath] ?? createI18nContext();
}
