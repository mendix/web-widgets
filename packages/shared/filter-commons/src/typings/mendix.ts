import { FilterCondition } from "mendix/filters";

export type FilterName = FilterCondition extends { name: infer Name } ? Name : never;
