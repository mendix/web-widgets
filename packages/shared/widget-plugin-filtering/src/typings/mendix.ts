import { FilterCondition } from "mendix/filters";
import { FnName } from "./type-utils";

export type FilterFunction = FnName<FilterCondition>;
