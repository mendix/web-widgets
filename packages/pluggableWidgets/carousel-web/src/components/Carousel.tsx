import { createElement, ReactNode, ReactElement, useCallback, useState } from "react";
import { GUID } from "mendix";
import classNames from "classnames";
import { SwiperOptions, A11y, Navigation, Pagination, EffectFade, Autoplay } from "swiper";
import { Swiper as ReactSwiper, SwiperClass, SwiperSlide } from "swiper/react";
import { PaginationOptions } from "swiper/types";

interface CarouselItem {
    id: GUID;
    content?: ReactNode;
}

export interface CarouselProps {
    id: string;
    pagination: boolean;
    loop: boolean;
    animation?: boolean;
    autoplay?: boolean;
    delay?: number;
    navigation: boolean;
    className: string;
    tabIndex?: number | undefined;
    onClick?: () => void;
    items: CarouselItem[];
}

export function Carousel(props: CarouselProps): ReactElement {
    const { items, pagination, loop, animation, autoplay, delay, navigation, className, tabIndex, id, onClick } = props;
    const [activeIndex, setActiveIndex] = useState<number>(0);

    const getSlideId = useCallback(
        (item: CarouselItem | undefined): string => {
            return `carousel-slide-${id}-${item?.id}`;
        },
        [id]
    );

    const paginationOptions: PaginationOptions | boolean = pagination && {
        type: "bullets",
        clickable: true,
        renderBullet: (index, className) =>
            `<span role="button" aria-controls="${getSlideId(
                items[index]
            )}" aria-label="Go to slide ${index}" class="${className}"></span>`
    };

    const options: SwiperOptions = {
        slidesPerView: 1,
        centeredSlides: true,
        loop,
        navigation,
        autoplay: autoplay ? { delay, stopOnLastSlide: true } : false,
        pagination: paginationOptions,
        ...(animation && {
            effect: "fade",
            fadeEffect: { crossFade: true }
        }),
        modules: [A11y, Navigation, Pagination, EffectFade, Autoplay],
        a11y: {
            enabled: true,
            slideRole: "listitem"
        }
    };

    const updateSwiperIndex = useCallback((swiper: SwiperClass) => {
        setActiveIndex(swiper.realIndex);
    }, []);

    return (
        <div className={classNames(className, "widget-carousel")} tabIndex={tabIndex}>
            <ReactSwiper
                onActiveIndexChange={updateSwiperIndex}
                wrapperTag={"ul"}
                {...options}
                onClick={onClick}
                onSwiper={updateSwiperIndex}
            >
                {items?.map((item, index) => (
                    <SwiperSlide tag={"li"} aria-hidden={index !== activeIndex} key={item.id} id={getSlideId(item)}>
                        {item.content}
                    </SwiperSlide>
                ))}
            </ReactSwiper>
        </div>
    );
}
