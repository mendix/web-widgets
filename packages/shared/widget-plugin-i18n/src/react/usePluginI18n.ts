import { useState } from "react";
import { I18nController } from "../controllers/I18nController";
import { PluginI18n } from "../types";

export function usePluginI18n(languageCode: string): PluginI18n | null {
    const [plugin, setPlugin] = useState<PluginI18n | null>(() => {
        new I18nController({
            languageCode,
            onPluginChange: plugin => setPlugin(plugin)
        });

        return null;
    });

    return plugin;
}
