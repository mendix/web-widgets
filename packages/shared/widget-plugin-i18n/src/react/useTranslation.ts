import { useContext } from "react";
import { PluginI18n } from "../types";
import { getGlobalI18nPluginContext } from "./context";

export function useTranslation(): PluginI18n | null {
    return useContext(getGlobalI18nPluginContext());
}
