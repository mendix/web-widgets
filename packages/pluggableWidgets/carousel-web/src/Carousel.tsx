import { createElement, ReactNode, useCallback, ReactElement, useId } from "react";
import { ValueStatus, GUID, ObjectItem } from "mendix";
import { executeAction } from "@mendix/widget-plugin-platform/framework/execute-action";
import { CarouselContainerProps } from "../typings/CarouselProps";
import { Carousel as CarouselComponent } from "./components/Carousel";
import loadingCircleSvg from "./ui/loading-circle.svg";
import classNames from "classnames";
import "./ui/Carousel.scss";

export function Carousel(props: CarouselContainerProps): ReactNode {
    const { showPagination, loop, tabIndex, navigation, animation, delay, autoplay } = props;
    const onClick = useCallback(() => executeAction(props.onClickAction), [props.onClickAction]);
    const id = useId();

    const renderCarousel = (): ReactElement => {
        return (
            <CarouselComponent
                id={id}
                className={props.class}
                tabIndex={tabIndex}
                pagination={showPagination}
                loop={loop}
                animation={animation}
                autoplay={autoplay}
                delay={delay}
                navigation={navigation}
                items={
                    props.dataSource?.items?.map((item: ObjectItem) => ({
                        id: item.id as GUID,
                        content: props.content?.get(item)
                    })) ?? []
                }
                onClick={onClick}
            />
        );
    };
    const renderLoading = (): ReactNode => {
        return (
            <div className={classNames(props.class, "widget-carousel")} tabIndex={tabIndex}>
                <img src={loadingCircleSvg} className="widget-carousel-loading-spinner" alt="" aria-hidden />
            </div>
        );
    };

    return props.dataSource?.status !== ValueStatus.Available ? renderLoading() : renderCarousel();
}
