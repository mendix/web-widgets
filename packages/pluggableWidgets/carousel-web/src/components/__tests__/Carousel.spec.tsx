import { createElement } from "react";
import { Carousel, CarouselProps } from "../Carousel";
import renderer from "react-test-renderer";
import { GUID } from "mendix";

beforeEach(() => {
    jest.resetAllMocks();
});
describe("Carousel", () => {
    const defaultCarouselProps: CarouselProps = {
        id: "Carousel",
        className: "",
        items: [
            { id: "1" as GUID, content: <div>test1</div> },
            { id: "2" as GUID, content: <div>test2</div> }
        ],
        pagination: true,
        animation: true,
        autoplay: true,
        delay: 3000,
        loop: true,
        navigation: true,
        onClick: () => jest.fn()
    };

    it("renders correctly", () => {
        const carousel = renderer.create(<Carousel {...defaultCarouselProps} />);

        expect(carousel).toMatchSnapshot();
    });
    it("renders correctly without pagination", () => {
        const carousel = renderer.create(<Carousel {...defaultCarouselProps} pagination={false} />);

        expect(carousel).toMatchSnapshot();
    });
    it("renders correctly without navigation", () => {
        const carousel = renderer.create(<Carousel {...defaultCarouselProps} navigation={false} />);

        expect(carousel).toMatchSnapshot();
    });
});
