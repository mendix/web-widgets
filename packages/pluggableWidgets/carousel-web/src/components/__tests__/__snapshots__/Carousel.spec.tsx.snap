// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders correctly 1`] = `
<div
  className="widget-carousel"
>
  <div
    className="swiper"
  >
    <ul
      className="swiper-wrapper"
    >
      <li
        aria-hidden={false}
        className="swiper-slide"
        id="carousel-slide-Carousel-1"
        onLoad={[Function]}
      >
        <div>
          test1
        </div>
      </li>
      <li
        aria-hidden={true}
        className="swiper-slide"
        id="carousel-slide-Carousel-2"
        onLoad={[Function]}
      >
        <div>
          test2
        </div>
      </li>
    </ul>
    <div
      className="swiper-button-prev"
    />
    <div
      className="swiper-button-next"
    />
    <div
      className="swiper-pagination"
    />
  </div>
</div>
`;

exports[`Carousel renders correctly without navigation 1`] = `
<div
  className="widget-carousel"
>
  <div
    className="swiper"
  >
    <ul
      className="swiper-wrapper"
    >
      <li
        aria-hidden={false}
        className="swiper-slide"
        id="carousel-slide-Carousel-1"
        onLoad={[Function]}
      >
        <div>
          test1
        </div>
      </li>
      <li
        aria-hidden={true}
        className="swiper-slide"
        id="carousel-slide-Carousel-2"
        onLoad={[Function]}
      >
        <div>
          test2
        </div>
      </li>
    </ul>
    <div
      className="swiper-pagination"
    />
  </div>
</div>
`;

exports[`Carousel renders correctly without pagination 1`] = `
<div
  className="widget-carousel"
>
  <div
    className="swiper"
  >
    <ul
      className="swiper-wrapper"
    >
      <li
        aria-hidden={false}
        className="swiper-slide"
        id="carousel-slide-Carousel-1"
        onLoad={[Function]}
      >
        <div>
          test1
        </div>
      </li>
      <li
        aria-hidden={true}
        className="swiper-slide"
        id="carousel-slide-Carousel-2"
        onLoad={[Function]}
      >
        <div>
          test2
        </div>
      </li>
    </ul>
    <div
      className="swiper-button-prev"
    />
    <div
      className="swiper-button-next"
    />
  </div>
</div>
`;
